src/cargo/ops/cargo_package.rs

```diff
@@ -258,67 +258,91 @@ fn build_ar_list(
+            }
+            Err(_) => {
+                // The file exists somewhere outside of the package.
+                let file_name = file_path.file_name().unwrap();
+                if result
+                    .iter()
+                    .any(|ar| ar.rel_path.file_name().unwrap() == file_name)
+                {
+                    ws.config().shell().warn(&format!(
+                        "{} `{}` appears to be a path outside of the package, \
                             but there is already a file named `{}` in the root of the package. \
                             The archived crate will contain the copy in the root of the package. \
-                            Update the license-file to point to the path relative \
+                            Update the {} to point to the path relative \
                             to the root of the package to remove this warning.",
-                            license_file,
-                            license_name.to_str().unwrap()
-                        ))?;
-                    } else {
-                        result.push(ArchiveFile {
-                            rel_path: PathBuf::from(license_name),
-                            rel_str: license_name.to_str().unwrap().to_string(),
-                            contents: FileContents::OnDisk(abs_license_path),
-                        });
-                    }
+                        label,
+                        file,
+                        file_name.to_str().unwrap(),
+                        label,
+                    ))?;
+                } else {
+                    result.push(ArchiveFile {
+                        rel_path: PathBuf::from(file_name),
+                        rel_str: file_name.to_str().unwrap().to_string(),
+                        contents: FileContents::OnDisk(abs_file_path),
+                    })
                 }
             }
-        } else {
-            let rel_msg = if license_path.is_absolute() {
+        }
+    } else {
+        // If a readme does not exist it is a hard error
+        if label != "readme" {
+            let rel_msg = if file_path.is_absolute() {
                 "".to_string()
             } else {
                 format!(" (relative to `{}`)", pkg.root().display())
             };
             ws.config().shell().warn(&format!(
-                "license-file `{}` does not appear to exist{}.\n\
-                Please update the license-file setting in the manifest at `{}`\n\
+                "{} `{}` does not appear to exist{}.\n\
+                Please update the {} setting in the manifest at `{}`\n\
                 This may become a hard error in the future.",
-                license_path.display(),
+                label,
+                file_path.display(),
                 rel_msg,
+                label,
                 pkg.manifest_path().display()
             ))?;
         }
     }
```

Comment:
imo this shouldn't be in the shared function.  If we end up wanting to reuse it, it could be a function on its own.

Reply:
I agree, I'm just not sure where to have the error message since we check if the file exists inside of it. We could maybe change the function signature to take in a `warn_not_exist: bool` and remove `if label != "readme"`

Reply:
Adding bools to a function like that is suspiciousm, that we might be making a god-function.

One idea is if we keep the joining / existence check out of the function, then we should be able to make this work.

Reply:
I did just that, let me know what you think.